entgql/plugin/plugin.go

```diff
@@ -0,0 +1,246 @@
+		schema *ast.Schema
+
+		relaySpec bool
+	}
+
+	// EntGQLPluginOption is a option for the EntGQL plugin
+	EntGQLPluginOption func(*EntGQL) error
+)
+
+var (
+	annotationName = entgql.Annotation{}.Name()
+	camel          = gen.Funcs["camel"].(func(string) string)
+
+	_ plugin.Plugin              = (*EntGQL)(nil)
+	_ plugin.EarlySourceInjector = (*EntGQL)(nil)
+	_ plugin.ConfigMutator       = (*EntGQL)(nil)
+)
+
+// WithRelaySpecification adds the Relay specification to the schema
+func WithRelaySpecification(relaySpec bool) EntGQLPluginOption {
+	return func(e *EntGQL) error {
+		e.relaySpec = relaySpec
+		return nil
+	}
+}
+
+// NewEntGQLPlugin creates a new EntGQL plugin
+func NewEntGQLPlugin(graph *gen.Graph, opts ...EntGQLPluginOption) (*EntGQL, error) {
+	nodes, err := entgql.FilterNodes(graph.Nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	e := &EntGQL{
+		graph: graph,
+		nodes: nodes,
+	}
+	for _, opt := range opts {
+		if err = opt(e); err != nil {
+			return nil, err
+		}
+	}
+
+	types, err := e.buildTypes()
+	if err != nil {
+		return nil, err
+	}
+
+	e.schema = &ast.Schema{
+		Types: types,
+	}
+	return e, nil
+}
+
+// Name implements the Plugin interface.
+func (*EntGQL) Name() string {
+	return "entgql"
+}
+
+// InjectSourceEarly implements the EarlySourceInjector interface.
+func (e *EntGQL) InjectSourceEarly() *ast.Source {
+	return nil
+	// return &ast.Source{
+	// 	Name:    "entgql.graphql",
+	// 	Input:   printSchema(e.schema),
+	// 	BuiltIn: false,
+	// }
+}
+
+func (e *EntGQL) buildTypes() (map[string]*ast.Definition, error) {
+	types := map[string]*ast.Definition{}
+	for _, node := range e.nodes {
+		ant, err := entgql.DecodeAnnotation(node.Annotations)
+		if err != nil {
+			return nil, err
+		}
+		if ant.Skip {
+			continue
+		}
+
+		fields, err := e.buildTypeFields(node)
+		if err != nil {
+			return nil, err
+		}
+
+		name := node.Name
+		if ant.Type != "" {
+			name = ant.Type
+		}
+
+		var interfaces []string
+		if e.relaySpec {
+			interfaces = append(interfaces, "Node")
+		}
+
+		types[name] = &ast.Definition{
+			Name:       name,
+			Kind:       ast.Object,
+			Fields:     fields,
+			Interfaces: interfaces,
+		}
+
+		// TODO(giautm): Added RelayConnection annotation check
+		if e.relaySpec {
+			defs, err := relayConnectionTypes(node)
+			if err != nil {
+				return nil, err
+			}
+
+			insertDefinitions(types, defs...)
+		}
+	}
+
+	return types, nil
+}
+
+func (e *EntGQL) buildTypeFields(t *gen.Type) (ast.FieldList, error) {
+	var fields ast.FieldList
+	if t.ID != nil {
+		f, err := e.typeField(t.ID, true)
+		if err != nil {
+			return nil, err
+		}
+		if f != nil {
+			fields = append(fields, f...)
+		}
+	}
+
+	for _, f := range t.Fields {
+		f, err := e.typeField(f, false)
+		if err != nil {
+			return nil, err
+		}
+		if f != nil {
+			fields = append(fields, f...)
+		}
+	}
+	return fields, nil
+}
+
+func (e *EntGQL) typeField(f *gen.Field, isID bool) ([]*ast.FieldDefinition, error) {
```

Comment:
Why do you return a slice here?

Reply:
```
	// TODO(giautm): support rename field
	// TODO(giautm): support mapping single field to multiple GQL fields
```
I have a few to do right here.

Reply:
> support mapping single field to multiple GQL fields

Can you explain in which case you'll do it?